src/cluster.c

```diff
@@ -4332,69 +4360,39 @@ void clusterReplyMultiBulkSlots(client *c) {
-            if (nodeFailed(node->slaves[i])) continue;
-            nested_elements++;
-        }
-
-        for (j = 0; j < CLUSTER_SLOTS; j++) {
-            int bit, i;
-
-            if ((bit = clusterNodeGetSlotBit(node,j)) != 0) {
-                if (start == -1) start = j;
+    int i = 0;
+    clusterNode *last_node = NULL;
+    int start_slot = 0;
+    int end_slot = 0;
+    int loop_range = CLUSTER_SLOTS + 1;
+    for (i = 0; i < loop_range; i++) {
+        if (NULL == last_node) {
+            if (CLUSTER_SLOTS != i && NULL != server.cluster->slots[i]) {
+                /* A new slots_range_start. */
+                last_node = server.cluster->slots[i];
+                start_slot = i;
+                end_slot = i;
             }
-            if (start != -1 && (!bit || j == CLUSTER_SLOTS-1)) {
-                addReplyArrayLen(c, nested_elements + 3); /* slots (2) + master addr (1). */
-
-                if (bit && j == CLUSTER_SLOTS-1) j++;
-
-                /* If slot exists in output map, add to it's list.
-                 * else, create a new output map for this slot */
-                if (start == j-1) {
-                    addReplyLongLong(c, start); /* only one slot; low==high */
-                    addReplyLongLong(c, start);
-                } else {
-                    addReplyLongLong(c, start); /* low */
-                    addReplyLongLong(c, j-1);   /* high */
-                }
-                start = -1;
-
-                /* First node reply position is always the master */
-                addReplyArrayLen(c, 3);
-                addReplyBulkCString(c, node->ip);
-                addReplyLongLong(c, node->port);
-                addReplyBulkCBuffer(c, node->name, CLUSTER_NAMELEN);
-
-                /* Remaining nodes in reply are replicas for slot range */
-                for (i = 0; i < node->numslaves; i++) {
-                    /* This loop is copy/pasted from clusterGenNodeDescription()
-                     * with modifications for per-slot node aggregation */
-                    if (nodeFailed(node->slaves[i])) continue;
-                    addReplyArrayLen(c, 3);
-                    addReplyBulkCString(c, node->slaves[i]->ip);
-                    addReplyLongLong(c, node->slaves[i]->port);
-                    addReplyBulkCBuffer(c, node->slaves[i]->name, CLUSTER_NAMELEN);
-                }
+            /* Skip if last node is null without a new slots_range_start. */
+        } else {
+            if (CLUSTER_SLOTS != i && NULL != server.cluster->slots[i] &&
+                server.cluster->slots[i] == last_node) {
+                /* Enlarge this slots range with last slots_range_start. */
+                end_slot = i;
+            } else {
+                /* Reset slots_range_start and make new reply for last slots_range, which
+                means a node. */
+                addNodeReplyForClusterSlot(c, last_node, start_slot, end_slot);
```

Comment:
I'm not sure end_slot is needed. The second value for this function is always i-1 or CLUSTER_SLOTS depending on the value of start_slot. That would remove the top half of this if statement.

Reply:
Yes, @KinWaiYuen you can refer to the implementation of `clusterGenNodesSlotsInfo`

Reply:
if all slots are assigned, none of node in `server.cluster->slots[x]` should be `NULL`. But some nodes may be `NULL` if slots are not assinged.